[Blackjack - Step2] Submit Deniz' mission

deniz-oezdemir opened the pull request:
Hi Eugine,

Thanks for reviewing step1. I have made the changes you recommended in step2. This PR is ready for review.

I have two questions regarding the same topic:
- Specifically, should I removed the result calculation and display logic from step1 completely since it is no longer used in step2? For now, I have only commented out the method calls in `Controller` but left the `displayResults` method in `OutputView` although it is not used anymore.
- Generally, is it best practice to delete all unused code?

Also, I am always looking to improve my code to be more Kotlin-idiomatic. I would appreciate any feedback on that.

Thanks in advance for your time. 🙏

Best, Deniz

petrabenczedh committed:
docs(readme): initial features

petrabenczedh committed:
init(blackjack): folder structure

petrabenczedh committed:
feat(blackjack): create Participant interface and Player class, test them

petrabenczedh committed:
feat(Controller): read player names and init players

petrabenczedh committed:
feat(OutputView): print player names

petrabenczedh committed:
feat(Card): create card class with rank and suit

petrabenczedh committed:
feat(Suit, Rank): create enums for card rank and suit

petrabenczedh committed:
feat(CardDeck): create CardDeck and test it

petrabenczedh committed:
feat(CardDeck, Participant): add draw card and hit functions

petrabenczedh committed:
feat(Dealer): new class

petrabenczedh committed:
feat(Controller): print handcards of players

petrabenczedh committed:
style(): format files with autoformatting

petrabenczedh committed:
feat(Controller): deal cards for players

petrabenczedh committed:
feat(HandCards, Player): calculate total sum and show if player is busted

petrabenczedh committed:
docs(README): update features

petrabenczedh committed:
test(Dealer): test drawing logic for dealer

petrabenczedh committed:
test(Dealer): test drawing logic for dealer

petrabenczedh committed:
feat(OutputView): display cards and totals summary

petrabenczedh committed:
refactor(HandCards): add type to properties

petrabenczedh committed:
docs(README)

petrabenczedh committed:
feat(Participant): add isBusted function

petrabenczedh committed:
feat(Players): add size property

petrabenczedh committed:
feat(ResultEvaluator): add class to calc results

petrabenczedh committed:
feat(OutputView): print results

petrabenczedh committed:
feat(ResultEvaluator): calculate results for players

petrabenczedh committed:
style(): minor changes

petrabenczedh committed:
style(): minor changes

petrabenczedh committed:
refactor(ResultEvaluator, HandCards): instead of properties use functions, operator overload in HandCards

petrabenczedh committed:
docs(README): questions

petrabenczedh committed:
docs(README): questions

petrabenczedh committed:
update(gitignore)

petrabenczedh committed:
update(gitignore)

petrabenczedh committed:
docs(README): questions

petrabenczedh committed:
refactor(Participant): use abstract class instead of interface for Participant

petrabenczedh committed:
refactor(ErrorMessage, PromptMessage): separate messages from core logic

petrabenczedh committed:
test(ResultEvaluator)

petrabenczedh committed:
style(OutputView)

petrabenczedh committed:
style(InputView)

petrabenczedh committed:
fix(OutputView): add new line before printing totals

petrabenczedh committed:
style(README):

deniz-oezdemir committed:
feat(gitignore): update gitignore

deniz-oezdemir committed:
Merge remote-tracking branch 'upstream/step1' into step1

deniz-oezdemir committed:
style(ResultEvaluator): remove commented pseudocode

deniz-oezdemir committed:
Update src/main/kotlin/blackjack/view/InputView.kt

Co-authored-by: Eugine Park <pkeugine@gmail.com>

deniz-oezdemir committed:
Update src/main/kotlin/blackjack/view/OutputView.kt

Co-authored-by: Eugine Park <pkeugine@gmail.com>

deniz-oezdemir committed:
refactor(OutputView): minimize `println()` calls
- use kotlin string templates and string multiline features to make fewer `println()` calls

deniz-oezdemir committed:
fix(DSLStudy): delete incomplete study test file

deniz-oezdemir committed:
fix(gitignore): delete wrong include for study test file from gitignore

deniz-oezdemir committed:
refactor(Rules): extract blackjack rule constants

deniz-oezdemir committed:
refactor(CardDeck): change deck to consist of 4 set of cards
- refactor tests to account for the new size of the card deck

deniz-oezdemir committed:
refactor(OutputView, ResultEvaluator): move result formatting to OutputView
- Add formatDealer() and formatPlayer() helpers
- displayResults() uses new helpers and raw string block
- Simplify toString() in DealerResult and PlayerResult for logging only

deniz-oezdemir committed:
style(TotalTest): add newline at the end of the file

deniz-oezdemir committed:
refactor(OutputView): centralize card formatting in OutputView
- replace model calls with OutputView helpers cardsToString and formatCard
- keep suit symbols mapping inside OutputView for consistent UI rendering

deniz-oezdemir committed:
style(input): move YES/NO constants to top of InputView object

deniz-oezdemir committed:
refactor(InputVIew): move input retry and validation handling to InputView
- clean up controller by delegating input retry and error printing to InputView
- add validation for name constraints when taking user input

deniz-oezdemir committed:
fix(OutputView): show card ranks as symbols (K, Q, J, A) instead of numeric values again

deniz-oezdemir committed:
fix(Controller, HandCards, InputVIew, OutputView, ResultsEvaluator): set methods private if possible

deniz-oezdemir committed:
refactor(Controller, Dealer): replace extension functions with internal service methods
- Remove Dealer.dealCards() and Players.dealCards() extension functions
- Add serviceDealer(), servicePlayers(), and dealCards() to Dealer
- Refactor tests according to above changes

deniz-oezdemir committed:
refactor(ResultEvaluatorTest): remove OutputView calls and rename test class

deniz-oezdemir committed:
build: remove unused kotlin("test") dependency

deniz-oezdemir committed:
refactor(Players): remove unused property `size`
- Apply autoformatting with ktlintFormat

deniz-oezdemir committed:
feature(DSL): Implement Kotlin DSL for Person creation with name, company, skills, and languages
- Add unit tests covering all DSL components and properties

deniz-oezdemir committed:
docs(readme): add features for step2 to feature list

deniz-oezdemir committed:
feature(InputView): read user input for players' bets
- Add new tests for player creation and refactor existing tests
- Remove unused methods in `Participant` class

deniz-oezdemir committed:
feature(Participants, OutputView): calculate and display earnings
- Add `settlePlayerAndDealer` method to `Participant` to calculate earnings for players and dealer
- Add tests for `settlePlayerAndDealer`
- Refactor OutputView to display final earnings

deniz-oezdemir committed:
refactor(Controller, OutputView): remove calls to calculation and display of results from step1

deniz-oezdemir committed:
feature(InputView): require betting amount to be multiple of 100

deniz-oezdemir committed:
style(Rules, ErrorMessage): replace magic numbers with values in Rules

deniz-oezdemir committed:
test(DeckTest): refactor of drawing more cards than deck contains

deniz-oezdemir committed:
refactor(DSL): support adding multiple soft/hard skills using vararg

deniz-oezdemir committed:
feature(PersonDSL, DSL): add DslMarker to use on builder methods

deniz-oezdemir committed:
style: apply autoformatting with ktlintFormat

deniz-oezdemir committed:
resolve conflicts by accepting version on step2

pkeugine commented on src/main/kotlin/blackjack/controller/Controller.kt:
```diff
@@ -30,9 +29,13 @@ class Controller {
+         * OutputView.displayResults(results)
+         */
+        dealer.settleRound()
+        OutputView.displayEarnings(dealer, players)
```
Thanks for showing me the difference!
The thing is, I can see the difference when I check the 'changed' code in GitHub:

<img width="1412" height="463" alt="Screenshot 2025-07-15 at 9 14 58 PM" src="https://github.com/user-attachments/assets/c7be5064-6fca-496b-aa4d-f21c87d14dbb" />

So I think we can safely remove this comment before this PR gets merged 😃

deniz-oezdemir replied:
Thanks for the hint, I will remove it. :-)

pkeugine commented on src/main/kotlin/blackjack/controller/Controller.kt:
```diff
@@ -42,7 +45,11 @@ class Controller {
-        return Players(playerNames.map { Player(it) })
+        val playerBets = InputView.readBets(playerNames)
+
+        val players = playerNames.zip(playerBets) { name, bet -> Player(name = name, bet = bet) }
```
Nice use of `zip` 👍

pkeugine commented on src/main/kotlin/blackjack/model/Participant.kt:
```diff
@@ -13,11 +15,18 @@ abstract class Participant {
-    fun cardsToString(): String {
-        return handCards.cards.joinToString()
-    }
+    fun settlePlayerAndDealer(dealer: Dealer) {
+        earning = when {
+            isBusted()                                                  -> -bet
+            dealer.isBusted()                                           -> +bet
+            handCards.isBlackjack() && dealer.handCards.isBlackjack()   -> 0
+            handCards.isBlackjack()                                     -> (bet * 1.5).toInt()
+            dealer.handCards.isBlackjack()                              -> -bet
+            handCards.total > dealer.handCards.total                    -> +bet
+            handCards.total < dealer.handCards.total                    -> -bet
+            else                                                        -> 0
+        }
```
Wow, I've never thought of using when like this; it looks pretty good.
A good way to make it easier to understand quickly might be adding some empty lines to group similar ones :  
```suggestion
    fun settlePlayerAndDealer(dealer: Dealer) {
        earning = when {
            isBusted()                                                  -> -bet
            dealer.isBusted()                                           -> +bet
            
            handCards.isBlackjack() && dealer.handCards.isBlackjack()   -> 0
            handCards.isBlackjack()                                     -> (bet * 1.5).toInt()
            dealer.handCards.isBlackjack()                              -> -bet
            
            handCards.total > dealer.handCards.total                    -> +bet
            handCards.total < dealer.handCards.total                    -> -bet
            
            else                                                        -> 0
        }
```

But this is just a small suggestion, so please do as you please 😉

deniz-oezdemir replied:
Yes, I agree. Grouping the cases makes it more readable. 😄

pkeugine commented on src/main/kotlin/study/DSL.kt:
```diff
@@ -11,16 +11,17 @@ data class Person(
     val company: String?,
     val skills: Skills,
-    val languages: List<Language>
+    val languages: List<Language>,
```
Nice addition of trailing comma 👍

deniz-oezdemir replied:
Thanks, but the credit goes to the auto-formatter here. 😉

pkeugine commented on src/main/kotlin/study/DSL.kt:
```diff
@@ -11,16 +11,17 @@ data class Person(
 
 data class Language(val name: String, val level: Int)
 
+@PersonDSL
```
Nice work of applying previous comment about `@DslMarker`! 💯

pkeugine commented on src/test/kotlin/blackjack/DeckTest.kt:
```diff
@@ -29,12 +26,11 @@ class DeckTest {
 
     @Test
-    fun `test draw more than deck`() {
+    fun `throws when drawing more cards than deck contains`() {
```
This test name definitely describes the test much better than the previous one 👍

pkeugine commented on src/test/kotlin/blackjack/ParticipantTest.kt:
```diff
@@ -0,0 +1,143 @@
+import org.junit.jupiter.api.Test
+
+class ParticipantTest {
+    private val deck = CardDeck()
```
Since this `deck` is only used in one test, why don't we keep it in the test scope?

Also, I think this would be a great time to think about tests, shared fixture, and test isolation.
If we were to use this `deck` in any other tests, though it might look okay for now, there will me moments where a test modifies this `deck` object and affect other tests. This will lead to unexpected test results, such as tests failing when they shouldn't or even worse, tests succeeding when they should fail.

Having tests create fixtures by themselves, or re-initiating shared fixtures are very important tasks when it comes to maintaining tests. It'll be great to think about this point in later missions :)

deniz-oezdemir replied:
Yes, the deck can definitely be instantiated in the respective test for this case. I will refactor it accordingly.

That's true, it will definitely not be good if the tests interact with each other because they use the same underlying data. For the future I will keep possible test interference in mind. Thanks for the hint! 🙏

pkeugine reviewed (changes requested):
Hello Deniz 👍 
The PR looks very solid, and I think we can wrap up this mission right away.
Most of the comments are just me verifying past comments and saying 'nice job'!
However, I thought you might want to edit some small parts,
so I'll keep it open, and merge it when you request a review once again.

If you have any questions or thoughts, feel free to DM me or leave comments here.
And as always,
Happy Hacking! 👏 👏 👏

deniz-oezdemir committed:
Update src/main/kotlin/blackjack/model/Participant.kt

- Group cases in `settlePlayerandDealer` for readability

Co-authored-by: Eugine Park <pkeugine@gmail.com>

deniz-oezdemir committed:
refactor(Controller, ParticipantTest): apply change requests from review
- Remove commented out step1 calculation calls
- Move card deck instantiation in the respective test to prevent test interference

deniz-oezdemir commented:
Hi @pkeugine ! 👋

Thanks so much for the kind words and thoughtful review! 😊

I went ahead and implemented the three changes you suggested, and everything should be good to go now.

Feel free to give it another look whenever you have a moment. It's ready for review.

Thanks again, and happy hacking to you too! 🙌

Best,
Deniz

pkeugine commented on README.md:
```diff
@@ -50,14 +50,17 @@ jason: Lose
   - [x] comma separated
```
**Q. Generally, is it best practice to delete all unused code?**

Personally, yes, I do think it is a good practice to remove unused code.
Even if we remove it, it'll still be in our git log, so it's not 100% gone from our work so we don't have to worry about it vanishing completely. Also, keeping unused code in the project tends to distract teams from focusing on what's actually important, and might confuse us by making us think 'this might be part of the logic we don't know', 'maybe it's important but we can't keep track of it'. When we need it again, it's not that hard to put it in.

pkeugine reviewed (approved):
Hello Deniz :)
Nice job on step 2, and I've answered the question you asked in the PR message.
It's my personal opinion based off of experience, so the answer might vary from developer to developer.

Anyways, the project looks like it's ready to get merged!
I hope you've learned a lot during this mission, and do well on your next mission as well.
Wish you all the luck, and as always,
Happy Hacking! 😄 😄 😄
